### PR TITLE
os/bluestore: configure rocksdb cache via bluestore_cache_kv_ratio

### PR DESCRIPTION
**Pull request by liewegas**
Also, enable the rocksdb row cache via rocksdb_cache_row_meta (default: .2)

Also, add rocksdb_cache_type = lru|clock option.  Note that 'clock' doesn't seem to work
in that it reports 'block_cache_name: LRUCache' when specified, and row_cache
comes back as None if I try to use it there.

### CODE REVIEW

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-08_20:27:41-rados-wip-sage-testing2-distro-basic-smithi/

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -1900,6 +1903,10 @@ class BlueStore : public ObjectStore,
   uint64_t kv_ios = 0;
   uint64_t kv_throttle_costs = 0;
 
+  float cache_meta_ratio = .1;   ///< cache ratio dedicated to metadata
+  float cache_kv_ratio = .1;     ///< cache ratio dedicated to kv (e.g., rocksdb)
+  float cache_data_ratio = .1;   ///< cache ratio dedicated to object data
+
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: Suggest to make them double and equal to 1 in total by default

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -3543,6 +3538,36 @@ void BlueStore::_set_blob_size()
            << std::dec << dendl;
 }
 
+int BlueStore::_set_cache_sizes()
+{
+  cache_meta_ratio = cct->_conf->bluestore_cache_meta_ratio;
+  cache_kv_ratio = cct->_conf->bluestore_cache_kv_ratio;
+  cache_data_ratio = 1.0 - cache_meta_ratio - cache_kv_ratio;
+
+  if (cache_meta_ratio <= 0 || cache_meta_ratio > 1.0) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Not sure that's a good approach to set values and then return an error.

File: src/common/config_opts.h [hunk 3]
```diff
@@ -1136,7 +1136,8 @@ OPTION(bluestore_cache_type, OPT_STR, "2q")   // lru, 2q
 OPTION(bluestore_2q_cache_kin_ratio, OPT_DOUBLE, .5)    // kin page slot size / max page slot size
 OPTION(bluestore_2q_cache_kout_ratio, OPT_DOUBLE, .5)   // number of kout page slot / total number of page slot
 OPTION(bluestore_cache_size, OPT_U64, 1024*1024*1024)
-OPTION(bluestore_cache_meta_ratio, OPT_DOUBLE, .9)
+OPTION(bluestore_cache_meta_ratio, OPT_DOUBLE, .5)
```
Comment (ifed01) [inline src/common/config_opts.h]: shouldn't we increase bluestore_cache_size with that default rocksdb_cache_size value and specify default ratios in a way that resulting disposition is similar to the one we have by default before. I.e. 128M to rocksdb, 900Mb to metadata and 100Mb to data cache. Just to not mess the feature and param tuning...

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: we error out in this case so the values aren't used.  this is mostly just less typing

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-11_18:26:58-rados-wip-sage-testing2-distro-basic-smithi/

**Pull request merged**